### PR TITLE
Light weight merkle tree implementation.

### PR DESCRIPTION
**Pull request by mrain**
## Description

closes: #141 

Gadget implementation not required.

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Targeted PR against correct branch (main)
- [x] Linked to GitHub issue with discussion and accepted design OR have an explanation in the PR that describes this work.
- [x] Wrote unit tests
- [x] Updated relevant documentation in the code
- [x] Added a relevant changelog entry to the `Pending` section in `CHANGELOG.md`
- [x] Re-reviewed `Files changed` in the GitHub PR explorer

### CODE REVIEW

File: primitives/src/merkle_tree/internal.rs [hunk 1]
```diff
@@ -226,6 +231,90 @@ where
...
+pub(crate) fn build_light_weight_tree_internal<E, H, I, Arity, T>(
+    height: usize,
+    elems: impl IntoIterator<Item = impl Borrow<E>>,
+) -> Result<(Box<MerkleNode<E, I, T>>, u64), PrimitivesError>
+where
+    E: Element,
+    H: DigestAlgorithm<E, I, T>,
+    I: Index + From<u64>,
+    Arity: Unsigned,
+    T: NodeValue,
```
Comment (chancharles92) [inline primitives/src/merkle_tree/internal.rs:244]: What's the difference from `NodeValue` and `Element`? Is that `Element` has a uid?

File: primitives/src/merkle_tree/internal.rs [hunk 2]
```diff
@@ -226,6 +231,90 @@ where
...
+{
+    let leaves: Vec<_> = elems.into_iter().collect();
+    let num_leaves = leaves.len() as u64;
+    let capacity = BigUint::from(Arity::to_u64()).pow(height as u32);
```
Comment (chancharles92) [inline primitives/src/merkle_tree/internal.rs:248]: How large the tree can be? Is it necessary to use `BigUint`?

File: primitives/src/merkle_tree/internal.rs [hunk 3]
```diff
@@ -576,7 +665,107 @@ where
...
+            MerkleNode::Leaf { .. } => Err(PrimitivesError::ParameterError(
+                "Incompatible merkle tree: index already occupied".to_string(),
+            )),
+            MerkleNode::ForgettenSubtree { .. } => Err(PrimitivesError::ParameterError(
+                "Given part of merkle tree is not in memory".to_string(),
+            )),
+        }
+    }
+
+    pub(crate) fn extend_and_forget_internal<H, Arity>(
```
Comment (chancharles92) [inline primitives/src/merkle_tree/internal.rs:691]: Maybe add some comments explaining the API's functionality and the usage of the API arguments.

**Review by chancharles92 — COMMENTED** [review]
👍

File: primitives/src/merkle_tree/internal.rs (hunk 1 above)
Comment (mrain) [reply primitives/src/merkle_tree/internal.rs:244]: `NodeValue` is the internal Merkle node value. `Element` is the type of the stored element.

File: primitives/src/merkle_tree/internal.rs (hunk 2 above)
Comment (mrain) [reply primitives/src/merkle_tree/internal.rs:248]: Good point. For an append-only light-weight Merkle tree `u64` or even `u32` suffices. Let me change it, with an additional range check.

Comment (chancharles92) [reply primitives/src/merkle_tree/internal.rs:248]: Thanks, let's still use u64 for safety and future-compatibility.

File: primitives/src/merkle_tree/light_weight.rs [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+    #[test]
+    fn test_light_mt_insertion() {
+        test_light_mt_insertion_helper::<Fq254>();
+        test_light_mt_insertion_helper::<Fq377>();
+        test_light_mt_insertion_helper::<Fq381>();
+    }
+
+    fn test_light_mt_insertion_helper<F: RescueParameter>() {
+        let mut mt = RescueLightWeightMerkleTree::<F>::from_elems(2, &[]).unwrap();
+        assert_eq!(mt.capacity(), BigUint::from(9u64));
```
Comment (chancharles92) [inline primitives/src/merkle_tree/light_weight.rs:99]: Is `capacity()` a BigUint?

File: primitives/src/merkle_tree/light_weight.rs [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+
+    #[test]
+    fn test_light_mt_builder() {
+        test_light_mt_builder_helper::<Fq254>();
+        test_light_mt_builder_helper::<Fq377>();
+        test_light_mt_builder_helper::<Fq381>();
+    }
+
+    fn test_light_mt_builder_helper<F: RescueParameter>() {
+        assert!(RescueLightWeightMerkleTree::<F>::from_elems(1, &[F::from(0u64); 3]).is_ok());
```
Comment (chancharles92) [inline primitives/src/merkle_tree/light_weight.rs:86]: Maybe make `3` a parameter (e.g. `let arity = RescueLightWeightMerkleTree::ARITY`)? Then replace the numbers below to be expressions related to `arity`.

The advantage is that when we change the setup of `RescueLightWeightMerkleTree`, the tests won't be broken.

File: primitives/src/merkle_tree/internal.rs [hunk 6]
```diff
@@ -226,6 +231,92 @@ where
...
+                    })
+                })
+                .collect_vec();
+            for i in 1..cur_nodes.len() - 1 {
+                cur_nodes[i] = Box::new(MerkleNode::ForgettenSubtree {
+                    value: cur_nodes[i].value(),
+                })
+            }
+        }
+        Ok((cur_nodes[0].clone(), num_leaves))
```
Comment (chancharles92) [inline primitives/src/merkle_tree/internal.rs:314]: What's the difference between `cur_nodes[0]` and `cur_nodes[0].clone()` here?

File: primitives/src/merkle_tree/internal.rs [hunk 7]
```diff
@@ -576,7 +673,109 @@ where
...
+                "Incompatible merkle tree: index already occupied".to_string(),
+            )),
+            MerkleNode::ForgettenSubtree { .. } => Err(PrimitivesError::ParameterError(
+                "Given part of merkle tree is not in memory".to_string(),
+            )),
+        }
+    }
+
+    /// Similar to [`extend_internal`], but this function will automatically
+    /// forget every leave except for the Merkle tree frontier.
```
Comment (chancharles92) [inline primitives/src/merkle_tree/internal.rs:700]: every leave -> every leaf

File: primitives/src/merkle_tree/light_weight.rs (hunk 4 above)
Comment (mrain) [reply primitives/src/merkle_tree/light_weight.rs:99]: Yes. It's from the trait definition of `MerkleTree`.

File: primitives/src/merkle_tree/internal.rs (hunk 6 above)
Comment (mrain) [reply primitives/src/merkle_tree/internal.rs:314]: `cur_nodes` is a local variable and it cannot be moved outside.

File: primitives/src/merkle_tree/light_weight.rs (hunk 5 above)
Comment (mrain) [reply primitives/src/merkle_tree/light_weight.rs:86]: Good point, will do.

File: primitives/src/merkle_tree/internal.rs (hunk 6 above)
Comment (chancharles92) [reply primitives/src/merkle_tree/internal.rs:314]: Ah right, it's a local vector; we can't just move the first element of it outside.

**Review by chancharles92 — APPROVED** [review]

**Pull request merged**